src/Storages/StorageS3.cpp

```diff
@@ -318,8 +354,18 @@ class StorageS3Source::KeysIterator::Impl : WithContext
 
                 for (const auto & key : keys)
                 {
-                    String path = fs::path(bucket) / key;
-                    String file = path.substr(path.find_last_of('/') + 1);
+                    const String path = fs::path(bucket) / key;
+                    const String file = path.substr(path.find_last_of('/') + 1);
+
+                    /// To avoid extra requests update total_size only if object_infos != nullptr
+                    /// (which means we eventually need this info anyway, so it should be ok to do it now).
+                    if (object_infos_)
+                    {
+                        auto key_info = S3::getObjectInfo(client_, bucket, key, version_id_, true, false);
+                        (*object_infos_)[path] = {.size = key_info.size, .last_modification_time = key_info.last_modification_time};
+                        total_size += key_info.size;
+                    }
```

Comment:
The same as above -- only filtered keys should be accounted. Also probably need to add `else` branch to calculate the total size. Is it really possible that `query` is `nullptr` or `virtual_header` is empty?